### PR TITLE
Closes #4067: Fennec logins migration

### LINKED ISSUE #4067: Fennec logins & password data import
Sibling of #4066.

Fennec will soon grow an ability to share logins/password data with trusted applications, using the same mechanism we rely on for FxA auto-login. See https://bugzilla.mozilla.org/show_bug.cgi?id=1572550

This work tracks getting logins information Fennec's ContentProvider, and passing it over to a-s for processing.
Depending on how Fennec shapes up, this may be either a path to a database file (preferred), or a bulk set of logins/passwords.

Relevant a-s ticket: https://github.com/mozilla/application-services/pull/1472

### PR DESCRIPTION
**Pull request by grigoryk**
---

### Pull Request checklist

- [x] **Quality**: This PR builds and passes detekt/ktlint checks (A pre-push hook is recommended)
- [x] **Tests**: This PR includes thorough tests or an explanation of why it does not
- [x] **Changelog**: This PR includes [a changelog entry](https://github.com/mozilla-mobile/android-components/blob/master/docs/changelog.md) or does not need one
- [x] **Accessibility**: The code in this PR follows [accessibility best practices](https://github.com/mozilla-mobile/shared-docs/blob/master/android/accessibility_guide.md) or does not include any user facing features

### After merge
- [ ] **Milestone**: Make sure issues closed by this pull request are added to the [milestone](https://github.com/mozilla-mobile/android-components/milestones) of the version currently in development.
- [ ] **Breaking Changes**: If this is a breaking change, please push a draft PR on [Reference Browser](https://github.com/mozilla-mobile/reference-browser) to address the breaking issues.

### CODE REVIEW

**Comment by grigoryk** [comment]
This PR doesn't expose `master password` as part of FennecMigrator's public API - but otherwise, it supports decryption of logins dbs protected with MP.

**Comment by grigoryk** [comment]
Fenix PR that enables logins migration: https://github.com/mozilla-mobile/fenix/pull/6930

File: components/support/migration/src/main/java/mozilla/components/support/migration/FennecLoginsMigration.kt [hunk 1]
```diff
@@ -0,0 +1,541 @@
...
+     * @param key4DbPath Path to key4 db.
+     * @return True if [masterPassword] is correct, false otherwise.
+     */
+    internal fun isMasterPasswordValid(masterPassword: String, key4DbPath: String): Boolean {
+        val db = SQLiteDatabase.openDatabase(key4DbPath, null, OPEN_READONLY)
+        return db.rawQuery("select item1, item2 from metadata where id = 'password'", null).use {
+            it.moveToNext()
+
+            val globalSalt = it.getBlob(it.getColumnIndexOrThrow("item1"))
+            val item2 = it.getBlob(it.getColumnIndexOrThrow("item2"))
```
Comment (csadilek) [inline components/support/migration/src/main/java/mozilla/components/support/migration/FennecLoginsMigration.kt]: Can we add the details from your comment https://github.com/mozilla-mobile/android-components/issues/4067#issuecomment-558424220 to describe what item2 is?

File: components/support/migration/src/main/java/mozilla/components/support/migration/FennecLoginsMigration.kt [hunk 2]
```diff
@@ -0,0 +1,541 @@
...
+                        timesUsed = timesUsed,
+                        timeCreated = timeCreated,
+                        timeLastUsed = timeLastUsed,
+                        timePasswordChanged = timePasswordChanged,
+                        usernameField = usernameField,
+                        passwordField = passwordField
+                    )
+                    records.add(fullRecord)
+                } catch (e: Exception) {
+                    crashReporter.submitCaughtException(e)
```
Comment (csadilek) [inline components/support/migration/src/main/java/mozilla/components/support/migration/FennecLoginsMigration.kt]: Here we'll send a caught exception to Sentry/Socorro for every incorrect record. How likely are incorrect records? Should we combine into one exception instead?

File: components/support/migration/src/main/java/mozilla/components/support/migration/FennecLoginsMigration.kt [hunk 3]
```diff
@@ -0,0 +1,541 @@
...
+        val globalSalt = getGlobalSalt(key4DbPath)
+        val db = SQLiteDatabase.openDatabase(key4DbPath, null, OPEN_READONLY)
+        val a11 = db.rawQuery("select a11, a102 from nssPrivate", null).use {
+            while (it.moveToNext()) {
+                val maybeRightKeyId = it.getBlob(it.getColumnIndexOrThrow("a102"))
+                if (maybeRightKeyId.contentEquals(keyId)) {
+                    return@use it.getBlob(it.getColumnIndexOrThrow("a11"))
+                }
+            }
+            null
```
Comment (csadilek) [inline components/support/migration/src/main/java/mozilla/components/support/migration/FennecLoginsMigration.kt]: nit: `return@use` here as well to make it easier to read?

File: components/support/migration/src/main/java/mozilla/components/support/migration/FennecLoginsMigration.kt [hunk 4]
```diff
@@ -0,0 +1,541 @@
...
+        return db.rawQuery("select item1 from metadata where id = 'password'", null).use {
+            it.moveToNext()
+            it.getBlob(it.getColumnIndexOrThrow("item1"))
+        }
+    }
+
+    /**
+     * Decrypts [ciphertext] using TDES.
+     */
+    private fun decryptTDes(key: ByteArray, iv: ByteArray, ciphertext: ByteArray): ByteArray {
```
Comment (csadilek) [inline components/support/migration/src/main/java/mozilla/components/support/migration/FennecLoginsMigration.kt]: nit `descryptTDES`?

File: components/support/migration/src/main/java/mozilla/components/support/migration/FennecLoginsMigration.kt (hunk 2 above)
Comment (csadilek) [reply components/support/migration/src/main/java/mozilla/components/support/migration/FennecLoginsMigration.kt]: Actually, yes, I think we should move this out of the loop. `decrypt` could throw a `NoSuchAlgorithmException` and then we'd be logging this a lot...

**Comment by grigoryk** [comment]
@csadilek addresses all of your feedback.

I've reached out to @mozkeeler for feedback, as well. Dana thinks this is basically fine (non-ideal as there are hand-rolled ASN.1 decoder bits, but alas). Main concern is that we should be careful about running this migration while Gecko is around. "NSS has historically had issues when anything else touches cert9.db/key4.db".
But, It will _probably_ be fine even if we overlap with NSS, somehow - "since it's read-only, the worst that should happen is some slowness".

**Comment by grigoryk** [comment]
bors r=csadilek

**Pull request merged**